csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -0,0 +1,154 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Ipc
+{
+    internal class ArrowStreamReaderImplementation : ArrowReaderImplementation
+    {
+        public Stream BaseStream { get; }
+        protected ArrayPool<byte> Buffers { get; }
+
+        public ArrowStreamReaderImplementation(Stream stream)
+        {
+            BaseStream = stream;
+            Buffers = ArrayPool<byte>.Create();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                BaseStream.Dispose();
+            }
+        }
+
+        public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken)
+        {
+            // TODO: Loop until a record batch is read.
+            cancellationToken.ThrowIfCancellationRequested();
+            return await ReadRecordBatchAsync(cancellationToken);
+        }
+
+        public override RecordBatch ReadNextRecordBatch()
+        {
+            throw new NotImplementedException();
+        }
+
+        protected async Task<RecordBatch> ReadRecordBatchAsync(CancellationToken cancellationToken = default)
+        {
+            await ReadSchemaAsync();
+
+            var bytesRead = 0;
+
+            byte[] lengthBuffer = null;
+            byte[] messageBuff = null;
+            byte[] bodyBuff = null;
+
+            try
+            {
+                // Get Length of record batch for message header.
+
+                lengthBuffer = Buffers.Rent(4);
+                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken);
+                var messageLength = BitConverter.ToInt32(lengthBuffer, 0);
+
+                if (messageLength == 0)
+                {
+                    //reached the end
+                    return null;
+                }
+
+                messageBuff = Buffers.Rent(messageLength);
+                bytesRead += await BaseStream.ReadAsync(messageBuff, 0, messageLength, cancellationToken);
+                var message = Flatbuf.Message.GetRootAsMessage(new FlatBuffers.ByteBuffer(messageBuff));
+
+                bodyBuff = Buffers.Rent((int)message.BodyLength);
+                var bodybb = new FlatBuffers.ByteBuffer(bodyBuff);
+                bytesRead += await BaseStream.ReadAsync(bodyBuff, 0, (int)message.BodyLength, cancellationToken);
+
+                return CreateArrowObjectFromMessage(message, bodybb);
+            }
+            finally
+            {
+                if (lengthBuffer != null)
+                {
+                    Buffers.Return(lengthBuffer);
+                }
```

Comment:
It'd be better to return the buffer as soon as we're done with it, so that it can be reused sooner, potentially even immediately when we try to rent the next one.

Reply:
👍 [ARROW-4503: ArrowStreamReader allocates and copies data excessively](https://issues.apache.org/jira/browse/ARROW-4503)